docs/internals/services.rst

```diff
@@ -1,6 +1,133 @@
+`dependency injection (DI) <https://wikipedia.org/wiki/Dependency_injection>`_,
+which is implemented in EF using `Microsoft.Extensions.DependencyInjection
+<https://docs.asp.net/en/latest/fundamentals/dependency-injection.html>`_.
+
+This article covers some fundamentals principles for understanding how EF uses
+services and DI.
+
+Categories of Services
+----------------------
+
+Services fall into one or more categories.
+
+Context services
+  Services that are related to a specific instance of  ``DbContext``. They
+  provide functionality for working with the user model and context options.
+
+Provider services
+  Provider-specific implementations of services. For example, SQLite uses
+  "provider services" to customize the behavior of SQL generation, migrations,
+  and file I/O as needed.
+
+Design-time services
+  Services used when a developer is creating an application. For example, EF
+  commands uses design-time services to execute migrations and code generation
+  (aka scaffolding).
+
+User services
+  A user can define custom services to interact with EF. These are written in
+  application code, not provider code. For example, users can provide an
+  implementation of ``IModelCustomizer`` for controlling how a model is created.
+
+
+Services Initialization
+-----------------------
+
+In order to be used by EF, all services must be registered in a service
+collection before or during ``DbContext`` initialization. Registration is done
+by adding components to into the ``ServiceCollection`` that EF will eventually
+use to create the final ``ServiceProvider``. By the time initialization is
+complete, ``DbContext`` has an instance of ``ServiceProvider`` that includes the
+collection of provider services, context services, and user services necessary
+for operation.
+
+.. note::
+  ``ServiceProvider`` is not to be confused with a "provider's services".
+
+Initialization Steps
+^^^^^^^^^^^^^^^^^^^^
+
+EF will initialize services in the following order. Services registered later
+can override or remove previously registered services.
+
+1. ``DbContext`` obtains an initial set of services. This initial set can be
+   internal or external. An external set must be found in order to use *user*
+   and *design-time services*. The initial set is discovered using these steps:
+  a. An **external** ``ServiceProvider`` can be passed in as a constructor
+     parameter.
+  b. If no constructor parameter is defined *and* ``DbContext`` is resolved from
+     a service provider (i.e. the context was registered with
+     ``AddDbContext<T>()``), then ``DbContextActivator`` will use the
+     **external** service provider from which the context was resolved.
+  c. If no external services can be found, ``DbContext`` creates an empty
+     **internal** ``ServiceProvider`` and calls ``.AddEntityFramework()`` to
+     add essential services.
+2. Extract any ``DbContextOptions`` relevant to the current DbContext type from
+   the initial service provider.
```

Comment:
Mention that this would normally be because AddDbContext was called on the ServiceCollection. Also mention that we create an empty one if none if found.